### ISSUE TITLE
Erroneous `deref_addrof` with immediate `&raw` pointer dereference

### ISSUE THREAD

**Issue by Supreeeme** [rust-lang/rust-clippy#15681]
### Summary

When immediately dereferencing a raw pointer created via `&raw const`/`&raw mut`, clippy will warn about immediately dereferencing a reference. This is wrong because I don't even have a reference in this situation, I have a raw pointer.

### Lint Name

deref_addrof

### Reproducer

I tried this code:

```rust
let a = Box::leak(Box::new(A { field: 5 })) as *const A;
let val = unsafe { *(&raw const ((*a).field)) };
```

I saw this happen:

```
warning: immediately dereferencing a reference
 --> src/main.rs:6:24
  |
6 |     let val = unsafe { *(&raw const ((*a).field)) };
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `((*a).field)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#deref_addrof
  = note: `#[warn(clippy::deref_addrof)]` on by default
```

I expected to see this happen:
No warning

### Version

```text
rustc 1.89.0 (29483883e 2025-08-04)
binary: rustc
commit-hash: 29483883eed69d5fb4db01964cdf2af4d86e9cb2
commit-date: 2025-08-04
host: x86_64-unknown-linux-gnu
release: 1.89.0
LLVM version: 20.1.7
```

**Comment by Jarcho** [comment]
This is technically not a reference, but the lint is still correct. The message really should use a term like "borrow expression" instead of "reference".

**Comment by Supreeeme** [comment]
How is the lint still correct? Raw pointers and references act differently.

**Comment by y21** [comment]
> How is the lint still correct? 

There should be no difference between `*&raw const (*struct).field` and just `(*struct).field` to my knowledge (in this case you end up copying the field out either way), hence the `help:` message suggests removing the no-op `*&raw const` part.

So only the `warning:` part of the diagnostic needs some rewording to not mention "references" as Jarcho said, and it's still correct to lint as it preserves semantics while still simplifying code

**Comment by Supreeeme** [comment]
The difference as I understand it is that `(*struct).field` will temporarily create a reference to `struct`, which can be undesirable when dealing with FFI, and `&raw const (*struct).field` will not.

**Comment by y21** [comment]
> The difference as I understand it is that `(*struct).field` will temporarily create a reference to struct

This doesn't sound right to me. `(*struct).field` merely creates a place and it depends on outer context whether a reference ultimately gets created for the final place.
- If you're writing `&(*struct).field` then there is a single reference being created to the final field only, not the whole struct
- If you write `let x = (*struct).field;`, then *no* reference is materialized at all, and the field's value is just copied out

This is necessarily the case for `#[repr(packed)]` structs to work correctly (soundly), otherwise the compiler would constantly insert unaligned references when accessing fields on them, something that you can do entirely in safe code.

If the field access involves going through a `Deref` impl, then you're right, it would create a reference for the `struct` due to the `Deref::deref` call, but I'm going to assume that's not the case here. There's a deny-by-default rustc lint that prevents this from happening.


In any case, adding `*&raw const` wouldn't change anything about whether `(*struct).field` inserts references or not.
Take a look at [this example](https://play.rust-lang.org/?version=stable&mode=release&edition=2024&gist=2265fd9b95d70dcdd4578171787f80a3). Going by your argument, if `(*struct).field` and `(&*struct).field` were equivalent because in both cases there's a reference materialized, then both cases should have UB.
However, Miri only reports UB for the first case. In the second case, there is no reference created, and as such has no UB.

**Comment by vigna** [comment]
I also think that with raw pointers this is wrong—unsafe reborrows can be now written concisely as `unsafe { &mut *(&raw mut x) }` and the lint should not warn about such instances.

**Comment by Jarcho** [comment]
That's a different case than what was first reported, but that should indeed not lint.
